GCC5 Build Failure [Darwin]

## Issue description
gcc5 and the dependant gfortran5 packages are failing as of 31 July
Failures: [ [gcc5](https://hydra.nixos.org/job/nixpkgs/trunk/gcc5.x86_64-darwin) [gfortran5](https://hydra.nixos.org/build/57627526#tabs-summary) ] [[hydra newly failing](https://hydra.nixos.org/eval/1380082?filter=gcc&compare=1379999&full=#tabs-now-fail)]


### Steps to reproduce
Building either package using `nix-build '<nixpkgs>' -A gcc5` or `nix-build '<nixpkgs>' -A gfortran5` fails.
## Technical details
```
../../gcc-5.4.0/libiberty/fibheap.c:151:7: warning: this function declaration is not a prototype [-Wstrict-prototypes]
../../gcc-5.4.0/libiberty/fibheap.c:220:30: error: use of undeclared identifier 'LONG_MIN'
  if (okey == key && okey != FIBHEAPKEY_MIN)
                             ^
../../gcc-5.4.0/libiberty/fibheap.c:38:24: note: expanded from macro 'FIBHEAPKEY_MIN'
#define FIBHEAPKEY_MIN  LONG_MIN
                        ^
../../gcc-5.4.0/libiberty/fibheap.c:261:36: error: use of undeclared identifier 'LONG_MIN'
  fibheap_replace_key (heap, node, FIBHEAPKEY_MIN);
                                   ^
```
* System: (MacOS 10.11.6 and MacOS 10.12.6, both nix-multiuser install)
* Nix version: (1.11.13)
* Nixpkgs version: (17.09pre-git)
* Sandboxing enabled: (relaxed)

--- a/pkgs/development/compilers/gcc/5/default.nix
+++ b/pkgs/development/compilers/gcc/5/default.nix
@@ -325,7 +325,13 @@ stdenv.mkDerivation ({
 
   NIX_LDFLAGS = stdenv.lib.optionalString  hostPlatform.isSunOS "-lm -ldl";
 
-  preConfigure = stdenv.lib.optionalString (hostPlatform.isSunOS && hostPlatform.is64bit) ''
+  preConfigure =
+    # Not sure why this is causing problems, now that the stdenv
+    # exports CPP=cpp the build fails with strange errors on darwin.
+    # https://github.com/NixOS/nixpkgs/issues/27889
+    stdenv.lib.optionalString stdenv.cc.isClang ''
+    unset CPP
+  '' + stdenv.lib.optionalString (hostPlatform.isSunOS && hostPlatform.is64bit) ''
     export NIX_LDFLAGS=`echo $NIX_LDFLAGS | sed -e s~$prefix/lib~$prefix/lib/amd64~g`
     export LDFLAGS_FOR_TARGET="-Wl,-rpath,$prefix/lib/amd64 $LDFLAGS_FOR_TARGET"
     export CXXFLAGS_FOR_TARGET="-Wl,-rpath,$prefix/lib/amd64 $CXXFLAGS_FOR_TARGET"

--- a/pkgs/development/tools/misc/gdb/default.nix
+++ b/pkgs/development/tools/misc/gdb/default.nix
@@ -66,6 +66,13 @@ stdenv.mkDerivation rec {
     ++ stdenv.lib.optional (!pythonSupport) "--without-python"
     ++ stdenv.lib.optional multitarget "--enable-targets=all";
 
+  preConfigure =
+    # Not sure why this is causing problems, now that the stdenv
+    # exports CPP=cpp the build fails with strange errors on darwin.
+    stdenv.lib.optionalString stdenv.cc.isClang ''
+      unset CPP
+    '';
+
   postInstall =
     '' # Remove Info files already provided by Binutils and other packages.
        rm -v $out/share/info/bfd.info

